Mission start splits

Hilimii opened the pull request:
This is a rough draft of the changes I wish to make. I believe they work as intended: I have tested Faust, Kings and F59 4 extensively to get them working with these new splits. Although, I need to thoroughly test back through most other missions and older features to see if my changes have altered anything. Until then, I present the code. Hopefully it's mistake free, but if you spot anything out of order, let me know!

Changes:
+ Lots of comments to organise and add clarity
+ New mission start splits setting (under campaign)
+ Mission start splits
+ Tunnel run function

Hilimii committed:
Mission Start Splits

Adds functionality to trigger splits at the beginning of each mission, once only.
Tunnel run split now has a dedicated function.
Added lots of new comments to organise things.
Rejigged the order of the 'split' event to accommodate new splits.

Hilimii committed:
Updates to Comments and Split event order

Adds numerous clarifying comments
Adds an additional 'if' to mission start splits, allowing logic to escape with doing much less work
Adjusts the position of Tunnel Split and Default Mission End Splits in the running order to ensure they trigger when desired.

Hilimii committed:
Logic Fix

Adjusts the order of split types and adds logic to Kings + Faust so that start splits can both trigger and not prevent other splits from triggering.
This was affecting most missions on campaign mode. I have learnt start splits need to be last in execution order to guarantee other splits can fire.

Hilimii committed:
Revamp split execution

Instead of using multiple ifs, mission start splits now refer to a table containing the relevant info within a for loop. Should massively improve efficiency.

jwfraustro commented:
I'm interested to see how this is coming along. Let me know when you're ready for me to take a look, this is looking pretty neat!

Hilimii committed:
Typo fix and testing

Fixed a typo that was affecting start splits on mission 2
removed section in 'start' that was removing vanilla campaign dictionary entries upon starting a F59 run
change to syntax on data remove from dictionary in 'Split' for start splits
Added prints to test data remove as mentioned above
Bump version

Hilimii commented on PW_Autosplitter.asl:
```diff
@@ -440,9 +426,14 @@ split
+                print(data.Value.Item1);
             }
         }
     }
```
I think it's almost ready! One thing I would appreciate help with is line 431. I'm trying to remove dictionary entries as they are used, so they aren't run over multiple times in the for loop used for start splits, but I don't think I'm getting the syntax quite right to do so. I've added prints to help see what the loop is doing in debug.

Hilimii commented on PW_Autosplitter.asl:
```diff
@@ -440,9 +426,14 @@ split
-            break;
+                print(data.Value.Item1);
+                return data.Value.Item2();
+                vars.StartSplitTable.Remove(data.Key); // This doesn't work, am I getting the syntax right?
```
I think it's almost ready! One thing I would appreciate help with is line 431. I'm trying to remove dictionary entries as they are used, so they aren't run over multiple times in the for loop used for start splits, but I don't think I'm getting the syntax quite right to do so. I've added prints to help see what the loop is doing in debug.

jwfraustro replied:
This might indicate a code smell that we should think about how we're doing this. However, generally, it's usually a bad idea to remove items from a collection while you're looping over it. Depending on the language, this can cause problems when the loop expects a value to be there, and finds that it has been removed on the next loop.

Hilimii replied:
I see! Thank you for warning me against this 😅

Hilimii commented:
I feel that this feature update is ready! The one issue with removing dictionary entries doesn't break anything, and if I can't get it working, I can easily remove the relevant line. The start split feature will work perfectly without it.

jwfraustro commented:
Awesome, I'll take a look!

jwfraustro commented:
I apologize for the delay in getting to this. I like a lot of the cleanup you've done here. 

I'm having a little bit of trouble understanding what the overall goal. Is it to just add individual start functions to each mission to make them more discrete? Or is it intended to allow IL time measurements during campaign runs (or is that just a handy side-effect)?

Hilimii commented:
That's no worries! The goal is the latter: To trigger a split at the start of each mission during a campaign run, which allows IL times to be measured or compared to during the run. See my most recent F59 to see it in action:
https://www.youtube.com/watch?v=68haqVRHwFQ

E.g. My PB for F59 4 Express Lane is 4:01. If, during a run, the time which has passed since I started F59 4 exceeds 4:01, I know that my pace on that mission is bad. But, if it's better, then I can edit out that segment in post as an IL run to submit. It's basically a big QoL change.

The reason that it's broken down per mission is so that repeat triggers of this type of split can be culled, otherwise, deaths would trigger these splits again. I based the current implementation off some code from the Prince of Persia autosplitter which I thought looked neat:
https://raw.githubusercontent.com/PoPRuns/PoP.AutoSplitters/main/pop_sot.asl

I know my implementation isn't the best, but I was limited by my knowledge 😅. If you can see any ways to neaten it up or improve the code, I am very much all ears. My hope is that breaking down per mission in a dictionary might be useful for future features that require a discrete approach.

Hilimii committed:
Simplification

Overhaul to massively simplify start split logic
- No longer relies on dictionary + functions on a discrete mission-by-mission basis
- Now watches a bool that is set to true until a mission ends, thereby preventing repeat triggers

Hilimii commented:
You'll be glad to know that I massively simplified the start splits logic so it doesn't rely on the dictionary anymore. It is also now generic in application, and does not work on a mission-by-mission basis. I was massively overcomplicating it before haha. 

I've kept in a small part of the dictionary, which includes GetLevelID results for each mission, because I thought it could be useful in the future.

jwfraustro commented:
Hi, sorry for not looking at this.... again. :( There's been a lot going on. I'll be able to take a look tomorrow, excited to see how it's shaping up!

Hilimii commented:
That's no worries! Life happens sometimes 😔 I been using the latest update for a while and haven't encountered a bug yet, I feel it's ready for release. Lemme know what you think :)

jwfraustro commented:
I had the chance to test it out and I'm very impressed-- it does what it says on the tin!

I spot-checked the campaign by skipping through some missions and it all seemed to work fine. If you've tested out the trickier ones for campaign / IL, I think this is good to go. I'm glad you found a way to cut down on the size of that data structure. 

I admit, I was initially confused at the double split when it first happened, and realized I had literally just read that it would do that in the tooltip...

I won't be doing any runs in the immediate future, but I'll keep it open for further testing whenever I happen to be on. 

Great job on this!
